projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+		if line[0] == '#' {
+			continue
+		}
+
+		// try to parse the player struct from the line and store it in the player struct
+		err := json.Unmarshal([]byte(line), &player)
+
+		// if we can't parse a player struct from the line then error but keep going
+		if err != nil {
+			continue
+		}
+
+		// add the player struct to the slice of players
+		*dataSlice = append(*dataSlice, player)
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in repeated JSON format")
+	}
+
+	return nil
+}
+
+// [3] CSV
+func parseCSV(file []byte, dataSlice *[]Player) error {
+	// because of the way we have tried to use file []byte as a parameter..
+	// it means we have to use bytes.NewReader on the file... this is JANKY
+	reader := csv.NewReader(bytes.NewReader(file))
+
+	// try to read the lines from the csv file
+	lines, err := reader.ReadAll()
+
+	// if we can't read the lines from the csv file then error
+	if err != nil {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	if len(lines) == 0 {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	// loop over each line
+	for _, line := range lines {
+
+		// (!!!) if we try to read one of the binary files it gives us this: [[��] [AyaPrisha����CharlieMargot]]
+		// so we need to add an additional check here:
+
+		// if the line does not have 2 values (highscore, name) then skip
+		// ((!) this is brittle if a line has valid data but random extra values etc)
+		if len(line) != 2 {
+			continue
+		}
+
+		// get the player name
+		name := line[0]
+
+		// the player has no name so skip
+		if len(name) < 1 {
+			continue
+		}
+
+		//  parse the string to get the score ((!) this is of type int not int32)
+		highScore, err := strconv.Atoi(line[1])
+
+		// if the player has no high score then skip
+		if err != nil {
+			continue
+		}
+
+		// add the player to the dataSlice
+		*dataSlice = append(*dataSlice, Player{Name: name, HighScore: int32(highScore)})
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	return nil
+}
+
+// [4] BINARY
+func parseBinary(file []byte, dataSlice *[]Player) error {
+	// each record contains exactly four bytes
+	// representing the score as a signed 32-bit integer (in the above endian format)
+	// then the name of the player stored in UTF-8
+	// which may not contain a null character, followed by a null terminating character
+
+	// --- DEBUG PRINT TABLE
+	// for index := 0; index < len(file); index++ {
+	// 	// print out the bytes in the slice in various formats
+	// 	fmt.Printf("slice index: %2d | Binary: %08b | Bytes (as an Integer): %3v | Hexadecimal: %2x | Decimal: %3d | ASCII : %c\n", index, file[index], file[index], file[index], file[index], file[index])
+	// }
+
+	// BIG ENDIAN (starts with fe ff) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
```

Comment:
Just to make sure we're clear: each two-hex-numbers (e.g. "fe") here represents one byte (a value 0-255), not two bytes

Reply:
I am glad you pointed this out
I need to go back over this because that is a fundamental misreading of mine